### ISSUE TITLE
Add explicit connection pooling options

### ISSUE THREAD

**Issue by emanguy** [kittinunf/Fuel#320]
I'm using Fuel in a high-concurrency environment, so due to UNIX's usage of sockets as files, I'm encountering issues where I have too many open sockets (due to the TIME_WAIT state of sockets after they're closed). This causes a state where I can no longer send HTTP requests due to the fact that I've hit my file-descriptor limit on the Java process.

The typical fix for this is connection pooling, but I don't know of a way to do this in Fuel. If options for this don't currently exist, I'd like to make a feature request for connection pooling in Fuel :)

**Comment by kittinunf** [comment]
I gotta admit that I am not an expert in this area. Do you an idea of a fix in mind? I would love to help you out if needed by discussing the potential fix (or feature) for this.

**Comment by immadolf** [comment]
So far, Fuel still has no connection pool, is it? Is there a guide can help me replace the default HttpClient with Apache HttpClient?

**Comment by iNoles** [comment]
@immadolf No, you can look into https://github.com/kittinunf/Fuel/blob/master/fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt

**Comment by emanguy** [comment]
@immadolf from what I've seen in the project it looks like you can implement the Client interface under the core library and set FuelManager.instance to it for it do do what you want.

I've used the Apache HTTPClient a bit, here's what's been working for me:
``` java
HTTPClientBuilder.create()
  .setMaxConnPerRoute(/* Number of persistent connections per route */)
  .setMaxConnTotal(/* Number of total persistent connections in the pool */)
  .build()
```
That should build a single thread-safe HTTP client. You'd just need to write your Apache HTTPClient implementation to parse the request and use appropriate HTTP methods/headers/etc. to conform to the Client protocol in Fuel.

**Comment by immadolf** [comment]
thks～
Actually,i did some work according to this mind.But it hasn't work now and
i am trying to find the problem.
Evan Rittenhouse <notifications@github.com>于2018年5月27日 周日10:35写道：
